pkg/resources/schemas/template_test.go

```diff
@@ -0,0 +1,375 @@
+			},
+		},
+		{
+			name: "Schema access attribute is updated",
+			setup: func(ctrl *gomock.Controller) testValues {
+				factory := schemafake.NewMockFactory(ctrl)
+				eventsReady := make(chan struct{})
+				baseSchemas := types.EmptyAPISchemas()
+				updateSchemas := types.EmptyAPISchemas()
+
+				testSchema := types.APISchema{
+					Schema: &v1schema.Schema{
+						ID:                "pod",
+						PluralName:        "pods",
+						CollectionMethods: []string{"GET"},
+						ResourceMethods:   []string{"GET"},
+					},
+				}
+				baseSchemas.AddSchema(testSchema)
+				// initial schemas
+				factory.EXPECT().Schemas(userInfo).Return(baseSchemas, nil)
+
+				// add access attribute
+				attributes.SetAccess(&testSchema, map[string]string{"List": "*"})
+				updateSchemas.AddSchema(testSchema)
+
+				// return updated schemas for the second request
+				factory.EXPECT().Schemas(userInfo).DoAndReturn(func(_ user.Info) (*types.APISchemas, error) {
+					// signal that schemas were requested
+					close(eventsReady)
+					return updateSchemas, nil
+				})
+
+				expectedEvents := []types.APIEvent{}
+				return testValues{expectedEvents, factory, eventsReady}
+			},
+		},
+	}
+
+	for i := range tests {
+		test := tests[i]
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+			// create new context for the test user
+			testCtx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			apiOp := &types.APIRequest{
+				Request: req.WithContext(request.WithUser(testCtx, userInfo)),
+			}
+
+			// create test factory
+			ctrl := gomock.NewController(t)
+			values := test.setup(ctrl)
+
+			// store onChange cb use to trigger the notifier that will be set in schemas.SetupWatcher(..)
+			var onChangeCB func()
+			values.mockFactory.EXPECT().OnChange(gomock.AssignableToTypeOf(testCtx), gomock.AssignableToTypeOf(onChangeCB)).
+				Do(func(_ context.Context, cb func()) {
+					onChangeCB = cb
+				})
+
+			baseSchemas := types.EmptyAPISchemas()
+
+			// create a new store and add it to baseSchemas
+			schemas.SetupWatcher(testCtx, baseSchemas, asl, values.mockFactory)
+			schema := baseSchemas.LookupSchema(resourceType)
+
+			// Start watching
+			resultChan, err := schema.Store.Watch(apiOp, nil, types.WatchRequest{})
+			assert.NoError(t, err, "Unexpected error starting Watch")
+
+			// wait for the store's go routines to start watching for onChange events
+			time.Sleep(setupTimeout)
+
+			// trigger watch notification that fetches new schemas
+			onChangeCB()
+
+			select {
+			case <-values.eventsReady:
+				// New schema was requested now we sleep to give time for watcher to send events
+				time.Sleep(setupTimeout)
+			case <-time.After(setupTimeout):
+				// When we continue here then the test will fail due to missing mock calls not being called.
+			}
+
+			// verify correct results are sent
+			hasExpectedResults(t, values.expectedChanges, resultChan, setupTimeout)
+		})
+	}
+}
```

Comment:
Would it be reasonable to add a test for Problem 2 checking for duplicate notifications?

Reply:
Yeah, I think those would be good tests for me to add.